### PR TITLE
Adding adaptive preconditioner functionality to Cantera (draft)

### PR DESCRIPTION
**Pull request by anthony-walker**
**Changes proposed in this pull request**

This code is adding preconditioning capabilities to Cantera/CVODEs solver interface for reactor networks. It is focused on a specific type of preconditioning but the code is being written with extensibility in mind. It adds a "Preconditioners" file to the numerics portion of cantera which contains a "PreconditionerBase" that is used to write future preconditioners. It also contains a class derived from this called "AdaptivePreconditioner" which is the aforementioned specific preconditioner. 

There are also a couple of functions added to specific reactors. IdealGasReactor is one example. Currently, I added a function to evaluate the energy equation which is also done inside of evalEqs but I wanted to avoid touching core Cantera code as much as possible. In this future, I would like to collaborate with the appropriate people to stream line this. 

**Checklist**

- [x] There is a clear use-case for this code change
- [x] The commit message has a short title & references relevant issues
- [x] Build passes (`scons build` & `scons test`) and unit tests address code coverage
- [x] The pull request is ready for review

### CODE REVIEW

File: include/cantera/numerics/AdaptivePreconditioner.h [hunk 1]
```diff
@@ -0,0 +1,177 @@
+/*
+Programmer: Anthony Walker
+This is the this file contains functions to adaptively precondition the sparse matrix class
+*/
```
Comment (bryanwweber) [inline include/cantera/numerics/AdaptivePreconditioner.h]: Can you use the standard header that we have for files in Cantera here?

File: include/cantera/numerics/AdaptivePreconditioner.h [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+          //!@param reactorStart an size_t providing the index location in which the state of the given reactor starts
+          virtual void reactorLevelSetup(IdealGasReactor* reactor, size_t reactorStart, double t, double* y, double* ydot, double* params);
+
+          //!Function used to complete individual reactor setups
+          //!@param reactor A IdealGasConstPressureReactor pointer
+          //!@param reactorStart an size_t providing the index location in which the state of the given reactor starts
+          virtual void reactorLevelSetup(ConstPressureReactor* reactor, size_t reactorStart, double t, double* y, double* ydot, double* params);
+      };
+} 
+#endif
```
Comment (bryanwweber) [inline include/cantera/numerics/AdaptivePreconditioner.h]: Please make sure that all your files end with a single blank line. If you're using VS Code, you configure two settings to trim final newlines and end a file with the newline to make sure you have one.

File: include/cantera/numerics/CVodesIntegrator.h [hunk 3]
```diff
@@ -10,7 +10,6 @@
 
 #include "cantera/numerics/Integrator.h"
 #include "cantera/base/ctexceptions.h"
-
```
Comment (bryanwweber) [inline include/cantera/numerics/CVodesIntegrator.h:13]: The changes in this file look spurious, can you remove them?

File: include/cantera/numerics/Integrator.h [hunk 4]
```diff
@@ -92,7 +93,8 @@ class Integrator
     /*!
      * @param probtype    Type of the problem
      */
-    virtual void setProblemType(int probtype) {
+    virtual void setProblemType(int probtype) 
+    {
```
Comment (bryanwweber) [inline include/cantera/numerics/Integrator.h]: I'm not sure what changed here, maybe line endings? Can you make sure you're using only "LF" line endings, not "CRLF"?

File: include/cantera/numerics/PreconditionerBase.h [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+#include "cantera/zeroD/FlowReactor.h"
+
+//Preconditioner type
+const int PRECONDITIONER_NOT_SET = 0;
+
+namespace Cantera//Making ASP apart of Cantera namespace
+{ 
+  class PreconditionerBase
+  { 
+    private:
```
Comment (bryanwweber) [inline include/cantera/numerics/PreconditionerBase.h]: This doesn't seem to be doing anything here.

File: src/numerics/CVodesIntegrator.cpp [hunk 6]
```diff
@@ -63,7 +62,7 @@ extern "C" {
         FuncEval* f = (FuncEval*) f_data;
         return f->eval_nothrow(t, NV_DATA_S(y), NV_DATA_S(ydot));
     }
-
+
```
Comment (bryanwweber) [inline src/numerics/CVodesIntegrator.cpp]: Please don't add any trailing whitespace to lines. You can also configure your editor to remove trailing whitespace.

File: src/numerics/CVodesIntegrator.cpp [hunk 7]
```diff
@@ -74,6 +73,29 @@ extern "C" {
         integrator->m_error_message = msg;
         integrator->m_error_message += "\n";
     }
+    
+    #if CT_SUNDIALS_VERSION>=30 //These functions only work with Sundials 3.0 or newer
```
Comment (bryanwweber) [inline src/numerics/CVodesIntegrator.cpp]: What happens if `CT_SUNDIALS_VERSION` is less than 3.0? Does the code fail with a useful error message?

File: src/numerics/CVodesIntegrator.cpp [hunk 8]
```diff
@@ -388,16 +410,36 @@ void CVodesIntegrator::applyOptions()
...
             #endif
         #endif
-    } else if (m_type == DIAG) {
+    } 
+    else if (m_type == DIAG) {
         CVDiag(m_cvode_mem);
-    } else if (m_type == GMRES) {
+    } 
+    else if (m_type == GMRES) 
+    {
```
Comment (bryanwweber) [inline src/numerics/CVodesIntegrator.cpp]: These changes appear spurious

**Review by bryanwweber — CHANGES_REQUESTED** [review]
Hi @anthony-walker Thanks for sharing your work so far! Since this is still in draft mode, I didn't make any comments about the implementation (well, except one I think about SUNDIALS). The comments are some small formatting things to watch out for as you continue working on this. Thanks!

File: src/numerics/CVodesIntegrator.cpp (hunk 7 above)
Comment (anthony-walker) [reply src/numerics/CVodesIntegrator.cpp]: This was updated to pass CI. It throws a `CanteraError` now. These functions are used within `CVodesIntegrator::applyOptions`. That is where the error is handled.

File: src/numerics/CVodesIntegrator.cpp (hunk 8 above)
Comment (anthony-walker) [reply src/numerics/CVodesIntegrator.cpp]: This was also changed already.

**Comment by ischoegl** [comment]
HI @anthony-walker ... I'm likewise happy to see this moving! To add to @bryanwweber 's preliminary review, I have some other minor formatting comments: `doublereal` should be `double`, and line lengths should be kept to 88 characters when possible (see [coding style guidelines](https://github.com/Cantera/cantera/blob/main/CONTRIBUTING.md)). Once the '(draft)' is removed from the title, I'll have additional comments.

**Comment by ischoegl** [comment]
PS: on a separate note, it looks like you're still on your fork's **main** branch for this PR (as noted elsewhere). I'm not sure that there is a simple fix that preserves the PR # (there *may* be), but it may make sense to switch to a feature branch on your fork sooner rather than later.

**Comment by anthony-walker** [comment]
@ischoegl I believe I would have to close the PR to change the branch but I will use a feature branch in the future.

**Comment by ischoegl** [comment]
> @ischoegl I believe I would have to close the PR to change the branch but I will use a feature branch in the future.

That's likely true, but it would spare you quite a few headaches whenever there is a need to rebase (as you've seen). Thus far, there isn't much attached to this PR, so re-opening won't lose too much information (i.e. what's above are mainly preliminary comments). Losing the commit record was unfortunate btw ... Choice is yours of course!

**Comment by anthony-walker** [comment]
Closing pull request to re-open from a feature branch.

**Comment by ischoegl** [comment]
Thank you, @anthony-walker ... for the records, this is continued on #1010.

**Pull request closed without merging**